cub/warp/warp_exchange.cuh

```diff
@@ -468,5 +468,89 @@ public:
   //@}  end member group
 };
 
+template <typename InputT, int ITEMS_PER_THREAD>
+class WarpExchangeRegister
```

Comment:
I'd prefer this to be a specialization instead of a separate API. You can take a look at cub/warp/specializations for reference. Would you have time to refactor your PR to provide something along the following lines?
```cpp
enum WarpExchangeAlgorithm
{
  WARP_EXCHANGE_SHMEM,
  WARP_EXCHANGE_SHUFFLE
};

template <typename InputT,
          int ITEMS_PER_THREAD,
          int LOGICAL_WARP_THREADS  = CUB_PTX_WARP_THREADS,
          int LEGACY_PTX_ARCH       = 0,
          WarpExchangeAlgorithm     = WarpExchangeAlgorithm::WARP_EXCHANGE_SHMEM>
class WarpExchange
// ...
```
This would allow new facility to expose the same API which might facilitate adoption, ease documentation and simplify tests. Regarding the tests, if you manage to introduce the algorithm parameter, would you mind adding the algorithm parameter to the warp exchange [test](`https://github.com/NVIDIA/cub/blob/main/test/catch2_test_warp_exchange.cu`)? You can take a look at the following [example](https://github.com/NVIDIA/cub/blob/6fa6d6e7715d43da8605c7b6766ca51035bdf4ec/test/catch2_test_block_load.cu#L167-L174) for reference.

Reply:
sure! For the general API, how much do you want the (partial) specialization to diverge? For instance, since this purely register based, the `__shared__ typename WarpExchangeT::TempStorage` "callback" doesn't make a lot of sense? Do you want me to omit it? I also don't see how the `ScatterToStriped` APIs can be implemented, since this would likely cause the compiler to not be able to infer the target lanes correctly, again dropping everything to local memory

Reply:
Please, take a look at cub/warp/specializations/warp_reduce_shfl.cuh for reference. The `WarpExchangeT::TempStorage` would have to be defined as `typedef NullType TempStorage`. Regarding the `ScatterToStriped` make sure to SFINAE it out in the `WarpExchange` if shuffle-based algorithm is required.

Reply:
@senior-zero just for posterity: do you want me to "customize" or partially specialize `class WarpExchange`? The former will require significant refactoring, since there's code like
```
  union _TempStorage
  {
    InputT items_shared[ITEMS_PER_TILE + PADDING_ITEMS];
  }; // union TempStorage

  /// Shared storage reference
  _TempStorage &temp_storage;
```
in `class WarpExchange`, which the warp version wouldn't provide?

Reply:
> Please, take a look at cub/warp/specializations/warp_reduce_shfl.cuh for reference.

I'd like the same approach as in warp reduce [implementation](https://github.com/NVIDIA/cub/blob/b7e042d43e752ef9b621b24d0322e831f2d162e9/cub/warp/warp_reduce.cuh#L159-L162). The temporary storage will be defined in both cases, it's just that in one of those it'll be an empty struct.